packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/props/cache.rs

```diff
@@ -0,0 +1,278 @@
+//! Allow for the caching of props (and state props and strings).
+
+use std::{borrow, cell::RefCell, collections::HashMap, rc::Rc};
+
+use crate::{
+    components::prelude::{GraphNode, PropCalcResult, PropValue},
+    graph::directed_graph::Taggable,
+    new_core::graph_node::GraphNodeLookup,
+    state::PropStatus,
+};
+
+/// Metadata stored along with a cached prop.
+#[derive(Debug, Clone)]
+struct CachedPropMeta {
+    /// Status of the prop indicating its initialization state and whether it is stale
+    pub status: PropStatus,
+    /// True if the value was set using the variable's default value.
+    /// Is typically used to lower the precedence of this prop's value
+    /// when calculating another prop, as it indicates that the document author
+    /// did not explicitly set the value
+    pub came_from_default: bool,
+    /// A counter that is incremented every time the value is set.
+    /// Used to compare with another counter to determine if the variable has been changed
+    /// since the other counter was last set.
+    pub change_counter: u32,
+}
+
+/// Store a prop's and other meta information needed to operate a cache
+/// that intelligently recomputes props when their dependencies change.
+#[derive(Debug)]
+struct CachedProp {
+    /// Cached value of the prop. Will be `None` if the prop hasn't been computed or is stale.
+    value: RefCell<Option<Rc<PropValue>>>,
```

Comment:
I'm still working out how this would be used. But, it seems to me that it'd be more useful with having an `Rc<RefCell<>>`. That way, one can store a copy of the cache somewhere and then have a handle to get its updated value. But maybe that is not the paradigm you are going for.

Having a `Rc` of an immutable value seems useful only to avoid cloning the `PropValue` or dealing with lifetimes/open Refs, which may be the reason you put the `Rc` here.

Reply:
Yeah. I did it for lifetime issues. Only one thing should ever be changing the cache. If someone is holding onto an old reference when the cached value is changed, that's fine. It's their duty to re-ask the cache for the value.